### PR TITLE
[RHCLOUD-47929] MCP tools: require user approval before write operations

### PR DESCRIPTION
**Pull request by dsova06**
## Link(s) to Jira
- [RHCLOUD-47929](https://issues.redhat.com/browse/RHCLOUD-47929)

## Description of Intent of Change(s)

All 23 MCP write tools now require explicit user approval before executing.
This prevents AI clients from accidentally creating, updating, or deleting
RBAC resources without human confirmation.

**How it works:**

1. When a write tool is called without a `confirmation_token`, the server
   returns a human-readable preview of the action (e.g., "This will create
   a new group 'Engineering Team'") and a single-use token.
2. The AI client presents this preview to the user and asks for approval.
3. If approved, the AI client calls the same tool again with the same
   arguments plus the `confirmation_token` to execute.

Delete operations show a `DESTRUCTIVE` prefix and irreversibility warning.

**Security properties:**
- Tokens are SHA-256 bound to tool name, arguments, and org_id
- Single-use (consumed on validation, preventing replay)
- Expire after configurable TTL (default 5 minutes)
- Cannot be transferred across tools, arguments, or organizations

**New settings:**
- `MCP_WRITE_CONFIRMATION` (default `True`) -- global toggle
- `MCP_WRITE_CONFIRMATION_TTL` (default `300`) -- token expiry in seconds

The confirmation protocol instructions are appended to the MCP `initialize`
response when both writes and confirmation are enabled, coaching AI clients
on the two-phase flow. Existing write test classes are updated to set
`MCP_WRITE_CONFIRMATION=False` so they continue to test tool CRUD directly.

## Local Testing

1. Enable write tools: `MCP_WRITE_ENABLED=True`
2. Call any write tool via the MCP endpoint without a `confirmation_token`:
   ```json
   {"jsonrpc": "2.0", "method": "tools/call", "id": 1,
    "params": {"name": "create_group", "arguments": {"name": "Test"}}}
   ```
3. Response should contain `confirmation_required: true` with a preview message and token.
4. Call again with the token to execute:
   ```json
   {"jsonrpc": "2.0", "method": "tools/call", "id": 2,
    "params": {"name": "create_group", "arguments": {"name": "Test", "confirmation_token": "<token>"}}}
   ```
5. Verify: reusing the token returns an error; changing arguments returns an error.
6. Set `MCP_WRITE_CONFIRMATION=False` and verify writes execute immediately.

Unit tests: `pipenv run tox -e py312-fast -- tests.management.test_mcp_views.MCPWriteConfirmationTests tests.management.test_mcp_views.MCPWritePreviewTests`

## Checklist
- [ ] if API spec changes are required, is the spec updated?
- [ ] are there any pre/post merge actions required? if so, document here.
- [x] are theses changes covered by unit tests?
- [ ] if warranted, are documentation changes accounted for?
- [x] does this require migration changes?
  - N/A -- no model changes
- [ ] is there known, direct impact to dependent teams/components?
  - AI clients consuming MCP write tools will need to implement the two-phase confirmation flow. The protocol is self-describing via the `instructions` field in the `initialize` response.

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
- [x] Input Validation
- [x] Output Encoding
- [x] Authentication and Password Management
- [x] Session Management
- [x] Access Control
- [x] Cryptographic Practices
- [x] Error Handling and Logging
- [x] Data Protection
- [x] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [x] General Coding Practices

[RHCLOUD-47929]: https://redhat.atlassian.net/browse/RHCLOUD-47929?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by Sourcery

Enforce a two-phase confirmation flow for all MCP write tools, requiring explicit user approval tokens before executing write operations, and expose this behavior via configuration and instructions.

New Features:
- Introduce a write confirmation protocol for MCP tools using single-use confirmation tokens tied to tool name, arguments, and organization.
- Add human-readable preview messages describing pending write operations before execution.
- Expose new MCP_WRITE_CONFIRMATION and MCP_WRITE_CONFIRMATION_TTL settings to control confirmation behavior and token expiry.

Enhancements:
- Extend MCP initialization instructions to document the write confirmation protocol when write tools and confirmation are enabled.

Tests:
- Add dedicated test suites for the write confirmation flow and preview message generation, and update existing MCP write tool tests to run with confirmation disabled.

### CODE REVIEW

**Review by dsova06 — COMMENTED** [review]
Clean feature with solid security properties -- SHA-256 binding, single-use tokens, TTL, org scoping. The test suite is thorough and the preview messages are clear. Two things stood out:

### Would be good to clean up

1. **Token consumed before validation completes** -- In `_validate_confirmation_token`, `django_cache.delete(key)` runs right after `get()`, before checking tool name, args_hash, and org_id. If an AI client accidentally sends a valid token to the wrong tool (e.g., `update_group` instead of `create_group`), the token is consumed and the user has to re-request confirmation from scratch. Since tokens are 128-bit UUID4 (brute-force infeasible), it would be safe to move the `delete()` after all validation checks pass -- that way honest client mistakes are recoverable without sacrificing security.

2. **No test for org_id mismatch** -- The suite covers invalid tokens, argument mismatch, and tool-name mismatch, but there's no test where a token issued for one org is presented by a different org. Worth adding alongside the other validation-path tests since it's a tenant isolation boundary.

### Something to think about

3. **`confirmation_token` lives outside tool input schemas** -- AI clients learn about `confirmation_token` through the `initialize` instructions, not through each tool's `inputSchema`. JSON Schema allows additional properties by default so this works, but if a client enforces strict validation on tool arguments, it would reject the extra parameter. If that surfaces in practice, one option would be adding `confirmation_token` as an optional field to write tools' schemas, or handling it as a protocol-level parameter outside of `arguments`.

Looks good otherwise, nice work.

File: rbac/management/mcp_views.py [hunk 1]
```diff
@@ -5031,6 +5050,116 @@ def _is_write_enabled() -> bool:
...
+    if tool_name == "add_principals_to_group":
+        principals = arguments.get("principals", [])
+        return f"This will add {len(principals)} principal(s) to group '{group_ref}'."
+    if tool_name == "add_roles_to_group":
+        roles = arguments.get("roles", [])
+        return f"This will assign {len(roles)} role(s) to group '{group_ref}'."
+    if tool_name in ("create_role_v1", "create_role"):
+        return f"This will create a new role '{name}'."
+    if tool_name == "create_role_bindings":
+        bindings = arguments.get("bindings", [])
```
[the remark this answers is not in the document]
Comment (dsova06) [reply rbac/management/mcp_views.py]: Fixed: all `len()` calls in `_generate_write_preview` now use `arguments.get("key") or []` to handle None/non-sequence values.

File: rbac/management/mcp_views.py [hunk 2]
```diff
@@ -5222,6 +5351,31 @@ def _handle_tools_call(request: HttpRequest, request_id: Any, params: dict[str,
             _record_metric(tool_name, "permission_denied")
             return _error_response(request_id, -32003, "Permission denied")
 
+    if config.write and _is_write_confirmation_enabled():
```
[the remark this answers is not in the document]
Comment (dsova06) [reply rbac/management/mcp_views.py:5406]: Fixed: `confirmation_token` is now always popped from arguments unconditionally before the write confirmation check, so it never leaks to tool functions regardless of whether confirmation is enabled.

File: tests/management/test_mcp_views.py [hunk 3]
```diff
@@ -7077,3 +7083,222 @@ def test_auth_check_runs_before_permission_check(self, _mock_v1, _mock_kessel):
         self.assertEqual(data["error"]["code"], -32000)
         _mock_kessel.assert_not_called()
         _mock_v1.assert_not_called()
+
+
+@override_settings(MCP_WRITE_ENABLED=True, MCP_WRITE_CONFIRMATION=True)
+class MCPWriteConfirmationTests(MCPToolTestMixin, IdentityRequest):
```
[the remark this answers is not in the document]
Comment (dsova06) [reply tests/management/test_mcp_views.py:7252]: Added: `test_token_rejected_for_different_org` creates a second tenant with a different org_id, issues a token under the original org, and asserts rejection with "different organization" when used under the second org.

File: tests/management/test_mcp_views.py [hunk 4]
```diff
@@ -7077,3 +7083,222 @@ def test_auth_check_runs_before_permission_check(self, _mock_v1, _mock_kessel):
...
+    def test_confirmation_token_not_passed_to_tool_function(self):
+        """The confirmation_token argument is stripped before calling the tool function."""
+        token = self._get_confirmation_token("create_group", {"name": "Token Strip Test"})
+
+        response2 = self._call_tool("create_group", {"name": "Token Strip Test", "confirmation_token": token})
+
+        data = response2.json()
+        self.assertFalse(data["result"]["isError"])
+        output = json.loads(data["result"]["content"][0]["text"])
+        self.assertEqual(output["name"], "Token Strip Test")
```
[the remark this answers is not in the document]
Comment (dsova06) [reply tests/management/test_mcp_views.py:7442]: Fixed: the test now uses `dataclasses.replace` + `patch.dict` to wrap the tool function with a `MagicMock`, then asserts `confirmation_token` is not in the kwargs passed to the tool.

**Comment by dsova06** [comment]
Addressed all Sourcery feedback plus self-review items:

- **Structured token field**: `confirmation_token` is now returned as a separate top-level field in the JSON payload alongside `message` and `confirmation_required`, so clients no longer need to parse the message string.
- **Hardcoded TTL**: Instructions now derive the expiry display from `MCP_WRITE_CONFIRMATION_TTL` settings instead of hardcoding "5 minutes".
- **Defensive `or []`**: All `len()` calls in `_generate_write_preview` use `arguments.get("key") or []` to handle None values.
- **Always pop token**: `confirmation_token` is unconditionally stripped from arguments before dispatch, regardless of confirmation feature flag.
- **Cross-org test**: Added `test_token_rejected_for_different_org` covering tenant isolation.
- **Token consumption**: Moved `django_cache.delete()` after all validation checks pass, so honest client mistakes don't consume the token.
- **Mock assertion**: Token stripping test now uses `dataclasses.replace` + `patch.dict` to verify `confirmation_token` is absent from tool kwargs.

@sourcery-ai resolve

**Comment by lpichler** [comment]
## Code Review Findings

### Issue 1: `confirmation_token` logged in argument keys (Minor / Informational) (Info)

**Location**: `rbac/management/mcp_views.py`, `_handle_tools_call`

The `confirmation_token` is popped from `arguments` at line ~5358, but the log statement at line ~5331 logs `list(arguments.keys())` before the pop occurs. This means `confirmation_token` will appear in log output as an argument key. While the token VALUE is not logged (only the key name), this leaks the existence of the confirmation flow in logs. This is very minor and likely intentional, but worth noting.

**Severity**: Informational -- no security impact since only the key name is logged, not the token value.

---

### 1. [Minor] No `confirmation_token` field in tool inputSchema (Info)

**Location**: `_handle_tools_list` / tool schema generation

When confirmation is enabled, write tools accept an optional `confirmation_token` argument, but this field is not reflected in the tool's `inputSchema` returned by `tools/list`. MCP clients that strictly validate arguments against the schema may refuse to send `confirmation_token`. The schema comes from FastMCP's `tool.inputSchema`, which is auto-generated from the Python function signature and won't include `confirmation_token` since it's popped from `arguments` before reaching the function.

**Severity**: Minor -- most MCP clients (including Claude) pass extra arguments freely, but spec-strict clients could reject it.

**Suggestion**: Consider injecting `confirmation_token` as an optional string property into the `inputSchema` of write tools when confirmation is enabled, or document that clients should pass it as an extra argument.

---

### 2. [Minor] Preview message for `remove_roles_from_group` lacks specificity (Info)

**Location**: `_generate_write_preview`, line handling `remove_roles_from_group`

```python
if tool_name == "remove_roles_from_group":
    return f"DESTRUCTIVE: This will remove role(s) from group '{group_ref}'. This is irreversible."
```

Unlike `add_roles_to_group` which shows the count of roles, `remove_roles_from_group` does not show which or how many roles will be removed. The tool accepts a `roles` argument (list of UUIDs), so the count could be shown.

**Severity**: Minor cosmetic -- the confirmation still works, the user just gets less detail.

File: rbac/management/mcp_views.py [hunk 5]
```diff
@@ -4981,12 +4983,32 @@ def delete(self, request: HttpRequest) -> HttpResponse:
...
+    "- The arguments in the second call must exactly match the first call."
+)
+
 
 def _build_mcp_instructions() -> str:
     """Build MCP server instructions based on current feature flags."""
     parts = [_MCP_INSTRUCTIONS_BASE, _MCP_INSTRUCTIONS_HONEST_CAVEATS, _MCP_INSTRUCTIONS_REMEDIATION]
     if _is_write_enabled():
         parts.append(_MCP_INSTRUCTIONS_WRITE_ACTIONS)
+        if _is_write_confirmation_enabled():
```
Comment (lpichler) [inline rbac/management/mcp_views.py:5048]: @dsova06 this will not work as lightspeed core doesn't support it. We needed to attach this stuff to each tool. More details in https://github.com/RedHatInsights/insights-rbac/pull/2928

**Review by lpichler — CHANGES_REQUESTED** [review]

File: rbac/management/mcp_views.py (hunk 5 above)
Comment (lpichler) [reply rbac/management/mcp_views.py:5048]: More details also here:

**Concern: confirmation UX relies on `initialize` instructions, which are not reliably followed by MCP clients**

The two-phase confirmation protocol correctly enforces safety **server-side** — the token validation and write blocking work regardless of the client. However, the human-approval UX layer depends on the AI client reading and acting on instructions injected via the MCP `initialize` response (`_MCP_INSTRUCTIONS_CONFIRMATION_PROTOCOL` in `_build_mcp_instructions()`, line ~5083).

In practice, we have already observed that `_MCP_INSTRUCTIONS_HONEST_CAVEATS` — added through the same `initialize` instructions mechanism — is not reliably followed by AI clients. The MCP `instructions` field is not guaranteed to be surfaced to the AI by all clients, and in our experience it has not been.

This means:
- The **server-side enforcement** (token system, blocking writes) works correctly ✅
- The **UX intent** (AI presents preview to human user, waits for explicit approval before re-calling) may be silently broken ❌

Without the `initialize` instructions working, the AI receives `{"confirmation_required": true, "message": "...", "confirmation_token": "..."}` but has no briefing on the expected protocol. It may auto-confirm by immediately re-calling with the token — defeating the human-in-the-loop purpose entirely.

**Suggestion**: The confirmation protocol instructions should be delivered through a channel that is reliably read by AI clients. Options:
1. Include the confirmation protocol description in each **write tool's `description`** field (returned in `tools/list`) — this is read by all compliant clients
2. Return the `confirmation_required` response as an MCP `isError: true` result with a structured message — some clients treat errors differently and are more likely to surface them to the user without auto-retrying
3. Do not rely on instructions for the UX; instead consider the `confirmation_required` response as the complete user-facing signal and test whether the target AI client handles it correctly without pre-briefing

Comment (dsova06) [reply rbac/management/mcp_views.py:5048]: Good catch -- you're right that initialize instructions don't reach Lightspeed Core (per findings in #2928).

Fixed: confirmation protocol instructions are now appended to each write tool's description in tools/list when confirmation is enabled. The suffix reads: "Write Confirmation: This tool requires user approval. The first call returns a preview and confirmation_token. Present the preview to the user. If approved, call again with the same arguments plus confirmation_token. NEVER auto-confirm."

The initialize instructions stay as a fallback for clients that do read them, but the tool descriptions are now the primary delivery channel -- matching the pattern established in #2928 for caveats.

Two new tests verify the suffix appears when confirmation is enabled and is omitted when disabled.

**Comment by lpichler** [comment]
@dsova06 it looks that CI failure is relevant to the PR

**Comment by dsova06** [comment]
@lpichler I ran both `tox -e py312-fast` (3582 tests) and `tox -e py312` (with coverage, matching CI) locally -- all tests pass. The Konflux logs only show `py312: FAIL code 1` without the actual test failure, so I can't tell which test failed in CI. Given the tests pass locally with both parallel modes, this looks like a Konflux infrastructure flake. Retriggering.

**Review by lpichler — APPROVED** [review]
